### PR TITLE
Dragen ica realignment pipeline

### PR DESCRIPTION
**Pull request by Alexander-Stuckey**
Pipeline for realigning CRAMs with Dragen in ICA.

Addition in `cpg_workflows` Dockerfile to install `icasdk`

There are a number of areas where this pipeline could be improved, but I wanted to get a working version out for review before I go on leave.

There are three empty stages at the moment: `GvcfMlrWithDragen`, `MonitorGvcfMlrWithDragen` and `CancelIcaPipeline`. The two MLR stages I am unlikely to get to before I go on leave. I may be able to work on the cancel stage before then. A cancel stage is needed, as this pipeline mostly makes API requests to Illumina's ICA platform, so hitting cancel in Hail Batch won't actually cancel any running pipeline in ICA.

There are a few places where checking for statuses could be more robust, again I just wanted something working in a PR before I go on leave.

### CODE REVIEW

File: cpg_workflows/stages/realign_genotype_with_dragen.py
Comment (MattWellie) [inline cpg_workflows/stages/realign_genotype_with_dragen.py:75]: There's a lot of use of `get_path_components_from_gcp_path`, the method is fine, but it feels like a long way round.

Instead of breaking into component parts and getting the "file" part, you could just use `sequencing_group.cram.name` for instance, which is much easier to read.

In general the Path objects have all the pieces you're looking for as instance attributes

```
cram_path_components['file'] == sequencing_group.cram.name
cram_path_components['bucket'] == sequencing_group.cram.drive
{cram_path_components["suffix"]}{cram_path_components["file"]} == sequencing_group.cram.blob
```

File: cpg_workflows/stages/realign_genotype_with_dragen.py
Comment (MattWellie) [inline cpg_workflows/stages/realign_genotype_with_dragen.py:324]: The performance of this 12-command block would be the same if it was all done as one multiline string block like this [example](https://github.com/populationgenomics/production-pipelines/blob/main/cpg_workflows/jobs/vqsr.py#L558). There's probably as much boilerplate as there is actual  commands so it's n t trivial to see the whole command sequence at a glance

File: cpg_workflows/stages/realign_genotype_with_dragen.py
Comment (MattWellie) [inline cpg_workflows/stages/realign_genotype_with_dragen.py:86]: As a general note having an explicit connection between the output of `expected_outputs` (generating the result paths) and the process that writes those files makes a lot of sense to me. Here you're building the paths to output files in `expected_outputs` from components, then building the same path elsewhere from components in a different way. 

The end result is that the outputs are created, but it's difficult to trace why/how they're created at the right paths due to the indirection.

Comment (MattWellie) [reply cpg_workflows/stages/realign_genotype_with_dragen.py:86]: This is probably the biggest deviation from every other stage in this pipeline framework. I'm guessing that's because the helper methods you've created (like `register_output_to_gcp`) require various parts of filename to use the google storage client, when you could use the Path object on the exact Path objects which are being created in `expected_ouputs`, e.g.

```
output_path = self.expected_outputs()

# do whatever to get the result
workflow_id = <some method call>

# open the Path for writing directly
with output_path.open('w') as handle:
    handle.write(workflow_id)
```

**Review by MattWellie — COMMENTED** [review]
It works, and I acknowledge that being the only target at this point, but there's quite a few deviations from the way (almost?) all our other stages have been written using this framework.

Some of that feels like a little bit of reinventing the wheel, or maybe just not finding some of the obscure helper methods which can attack the same problems using pretty tight syntax

File: cpg_workflows/stages/realign_genotype_with_dragen.py
Comment (MattWellie) [reply cpg_workflows/stages/realign_genotype_with_dragen.py:75]: Cloudpathlib (the result of `to_path('gs://something')`, returned by most of our helper methods, and PathLib have the same interface, so this would be a decent doc for all the various path elements these Path objects will have available as properties:

https://docs.python.org/3/library/pathlib.html#methods-and-properties

File: cpg_workflows/stages/realign_genotype_with_dragen.py
Comment (Alexander-Stuckey) [reply cpg_workflows/stages/realign_genotype_with_dragen.py:324]: Can do that, it's much cleaner

File: cpg_workflows/stages/realign_genotype_with_dragen.py
Comment (Alexander-Stuckey) [reply cpg_workflows/stages/realign_genotype_with_dragen.py:50]: Makes sense

Comment (Alexander-Stuckey) [reply cpg_workflows/stages/realign_genotype_with_dragen.py:50]: Ah, except that I need them as inputs for a couple of other stages, so not sure what the practice here is?

File: cpg_workflows/stages/dragen_ica/ica_utils.py
Comment (Alexander-Stuckey) [reply cpg_workflows/stages/dragen_ica/ica_utils.py:32]: Done

File: cpg_workflows/stages/realign_genotype_with_dragen.py
Comment (Alexander-Stuckey) [reply cpg_workflows/stages/realign_genotype_with_dragen.py:36]: Neat, that's much nicer

File: Dockerfile
Comment (Alexander-Stuckey) [reply Dockerfile:11]: Moved `icasdk` install to beginning of Dockerfile

File: configs/defaults/realign_genotype_with_dragen.toml
Comment (Alexander-Stuckey) [reply configs/defaults/realign_genotype_with_dragen.toml:31]: Can do

File: cpg_workflows/stages/realign_genotype_with_dragen.py
Comment (Alexander-Stuckey) [reply cpg_workflows/stages/realign_genotype_with_dragen.py:71]: I'll give this a try, I think it was having some trouble with `-test` earlier, which is why I had swapped to the current syntax. However, that problem seems to have reoccurred when using `bioheart-test` instead of `fewgenomes-test` so 🤷

File: cpg_workflows/stages/realign_genotype_with_dragen.py
Comment (MattWellie) [reply cpg_workflows/stages/realign_genotype_with_dragen.py:45]: yeah no worries that's cool. Answer iss..... depends how you want to access them. The analysis_keys in the `@stage` are only relevant for the metamist registration. If you want to pull from Metamist later, they have to be in there.

If you want to pull them directly from the previous stage as inputs the `inputs.as_dict(self, Stage_Name)` (or whatever the syntax is) will always get you the full result of the expected_outputs method, the stage decorator isn't involved

**Comment by MattWellie** [comment]
Just cross-posting this PR: https://github.com/populationgenomics/production-pipelines/pull/1068. The fix adopted in this PR was the one from #1063. #1068 is a more succinct fix for the `-test` suffix/`dataset.prefix()` issue, which leaves the metamist.py script unchanged.

I've made [this](https://github.com/populationgenomics/production-pipelines/pull/1069) stacked on top of this branch to update to the better version of the fix, and check that the fix works as intended before adopting.

**Review by MattWellie — COMMENTED** [review]
Looks good! There's a stacked PR to prevent this merge to main circumventing #1068, but this looks like a ruddy good start

File: cpg_workflows/stages/dragen_ica/monitor_align_genotype_with_dragen.py
Comment (Alexander-Stuckey) [reply cpg_workflows/stages/dragen_ica/monitor_align_genotype_with_dragen.py:50]: Yeah for sure. This was again just something to get working in case I had to disappear on leave

Comment (Alexander-Stuckey) [reply cpg_workflows/stages/dragen_ica/monitor_align_genotype_with_dragen.py]: Cleaned this up now, it raises an explicit exception if you cancel the pipeline (in ICA or otherwise), and otherwise informs you to check ICA for any other reason that the pipeline failed.

File: cpg_workflows/stages/realign_genotype_with_dragen.py
Comment (Alexander-Stuckey) [reply cpg_workflows/stages/realign_genotype_with_dragen.py]: I believe I've resolved all of these, except where strictly necessary

File: cpg_workflows/stages/realign_genotype_with_dragen.py
Comment (MattWellie) [inline cpg_workflows/stages/realign_genotype_with_dragen.py:87]: just out of curiosity is this result being kept as a `{key: result_ID}` dictionary because it needs to be passed that way in future? If the json'ification isn't important I would have assumed just returning/storing the ID in isolation would be easier (not needing JQ/other to read back later)

File: cpg_workflows/stages/realign_genotype_with_dragen.py
Comment (MattWellie) [inline cpg_workflows/stages/realign_genotype_with_dragen.py:128]:
```suggestion
        cram_group = get_batch().read_input_group(
            cram=str(sequencing_group.cram.path), 
            crai=str(sequencing_group.cram.index_path),
        )
        
        # Check if the CRAM already exists in ICA before uploading. If it exists, just return the ID for the CRAM and CRAI
        upload_job.command(
            f"""
            {ICA_CLI_SETUP}
            cram_status=$(icav2 projectdata list --parent-folder /{bucket}/{upload_folder}/{sequencing_group.name}/ --data-type FILE --file-name {sequencing_group.name}.cram --match-mode EXACT -o json | jq -r '.items[].id.details.status')
            if [[ $cram_status != "AVAILABLE" ]]
            then
                icav2 projectdata upload {cram_group.cram} /{bucket}/{upload_folder}/{sequencing_group.name}/
                icav2 projectdata upload {cram_group.crai} /{bucket}/{upload_folder}/{sequencing_group.name}/
            fi
```

just a suggestion to delegate localisation to hail batch, which would make it easier to drop the gcloud cli in this job image later

Comment (MattWellie) [reply cpg_workflows/stages/realign_genotype_with_dragen.py:128]: This diff is a bit messy given that it includes the unchanged lines in between. It's the read_input_group to create the local files from cram/crai, and dropping the two gcloud cp lines

File: cpg_workflows/stages/dragen_ica/run_align_genotype_with_dragen.py
Comment (MattWellie) [inline cpg_workflows/stages/dragen_ica/run_align_genotype_with_dragen.py:91]: If I'm being a picky bugger, seeing as so many of these methods are all called `run`, a docstring explanation of what this `run` runs would be cool. In the job call it's clearer as `run_align_genotype_with_dragen.run` but in this review I keep bouncing up to the file name to know what the method does

File: cpg_workflows/stages/realign_genotype_with_dragen.py
Comment (MattWellie) [inline cpg_workflows/stages/realign_genotype_with_dragen.py:320]: Aha, this the corresponding reading in of that JSON output. Seems fine, not sure if `$(cat analysis_id)` would be less work, but you might need to worry about line endings. 

Basically I've never gotten used to JQ, it scares me.

**Review by MattWellie — APPROVED** [review]
LGTMatt

File: cpg_workflows/stages/realign_genotype_with_dragen.py
Comment (Alexander-Stuckey) [reply cpg_workflows/stages/realign_genotype_with_dragen.py:87]: Yeah that'd be a better solution. Earlier this method had multiple key: value pairs, so json did make sense, but that changed once I had to redo the upload method. Will change

Comment (Alexander-Stuckey) [reply cpg_workflows/stages/realign_genotype_with_dragen.py:87]: I will look into changing this (as that would be better) in a new branch, because it'd be good to get something working merged in first.

File: cpg_workflows/stages/realign_genotype_with_dragen.py
Comment (Alexander-Stuckey) [reply cpg_workflows/stages/realign_genotype_with_dragen.py:128]: So I know that it doesn't take a huge amount of time to load data from GCP to the VM, but the current solution only does that if the CRAM is not already present in ICA, instead of always. Maybe I'm solving for a problem that won't really come up, but 🤷

File: cpg_workflows/stages/dragen_ica/run_align_genotype_with_dragen.py
Comment (Alexander-Stuckey) [reply cpg_workflows/stages/dragen_ica/run_align_genotype_with_dragen.py:91]: Yeah can do this

**Comment by jmarshall** [comment]
I wish you would have announced that you were planning to merge this imminently and put out a final call for review, either here on the PR or on slack. You have asked a number of people to review this, many of whom have may have done so if it had been clear that you now considered this ready for merging.

I have two comments, but as they are somewhat process-related I had been leaving them until I had time to do a more useful review of the PR's actual content as well.

1. Because everything is imported in _main.py_, in this unified codebase the use of `icasdk` will lead to trouble for other workflows and local dry-running:

```
File "/Users/johnm/production-pipelines/cpg_workflows/stages/dragen_ica/cancel_ica_pipeline_run.py", line 6, in <module>
    import icasdk
ModuleNotFoundError: No module named 'icasdk'
```

This may be annoying to work around in your source files, but it's probably not reasonable to expect all driver images and local laptops to install this SDK. Perhaps you could just do something like `try: import icasdk; except ImportError: pass # can't actually use this workflow then`.

2. Merging this PR has applied a variant of the `-test` fixes to _cpg_workflows/inputs.py_. This work should have been left to PR #1068 and this PR rebased on top of that / merged onto from **main** once the full fix for that had been merged to **main**.

**Comment by Alexander-Stuckey** [comment]
> I wish you would have announced that you were planning to merge this imminently and put out a final call for review, either here on the PR or on slack. You have asked a number of people to review this, many of whom have may have done so if it had been clear that you now considered this ready for merging.

This is fair, yet I will point out that this PR was marked ready for review last week, and only had reviews from one person in that time. Given that there had been no other apparent interest in that time, I had made the assumption that nobody else was wanting to review it.

Next time I will tag people for a final round of review before merging.


> I have two comments, but as they are somewhat process-related I had been leaving them until I had time to do a more useful review of the PR's actual content as well.
> 
> 1. Because everything is imported in _main.py_, in this unified codebase the use of `icasdk` will lead to trouble for other workflows and local dry-running:
> 
> ```
> File "/Users/johnm/production-pipelines/cpg_workflows/stages/dragen_ica/cancel_ica_pipeline_run.py", line 6, in <module>
>     import icasdk
> ModuleNotFoundError: No module named 'icasdk'
> ```

This is a very good solution, I will make a quick PR to ad this in so that this no longer breaks for others that have different workflows / images.

> This may be annoying to work around in your source files, but it's probably not reasonable to expect all driver images and local laptops to install this SDK. Perhaps you could just do something like `try: import icasdk; except ImportError: pass # can't actually use this workflow then`.



> 2. Merging this PR has applied a variant of the `-test` fixes to _cpg_workflows/inputs.py_. This work should have been left to PR [Removing test suffix from dataset names for multicohorts. #1068](https://github.com/populationgenomics/production-pipelines/pull/1068) and this PR rebased on top of that / merged onto from **main** once the full fix for that had been merged to **main**.

I apologise for this one, I'd used it for testing and forgotten to roll it back before merging, like I had for some other changes that were only in for testing. Would you like me to remove this, so that [Removing test suffix from dataset names for multicohorts. #1068](https://github.com/populationgenomics/production-pipelines/pull/1068) can still proceed properly?

**Comment by jmarshall** [comment]
Was adding `icasdk` to the driver image also part of just stopping this import from breaking for people running other workflows, or was it intentional and significant to add this SDK to the default driver image?

If the former, reverting this change to _Dockerfile_ would also be appreciated.

**Pull request merged**
